### PR TITLE
compute: support BackendServices

### PR DESCRIPTION
**Pull request by stephenplusplus**
For #1073 

This implements Backend Services, known here as "services".
#### To Dos
- [x] Tests
  - [x] System
  - [x] Unit
    - [x] compute/index
    - [x] compute/service

``` js
var service = compute.service('service-name');

service.getHealth({
  name: 'instance-group-name',
  zone: 'us-central1-a'
}, function(err, status) {
  status = [
    {
      ipAddress: '...',
      instance: '...',
      healthState: '...',
      port: '...'
    }
  ]
});
```

### CODE REVIEW

**Comment by stephenplusplus** [comment]
@callmehiphop I'm going to start on the unit tests, but let me know if you catch anything weird with the code.

**Comment by stephenplusplus** [comment]
<kbd>don't merge</kbd> lifted!

File: lib/compute/index.js [hunk 1]
```diff
@@ -306,6 +312,70 @@ Compute.prototype.createNetwork = function(name, config, callback) {
...
+ *
+ * @resource [Backend Services Overview]{@link https://cloud.google.com/compute/docs/load-balancing/http/backend-service}
+ * @resource [BackendServices: insert API Documentation]{@link https://cloud.google.com/compute/docs/reference/v1/backendServices/insert}
+ *
+ * @param {string} name - Name of the backend service.
+ * @param {object} config - See a
+ *     [BackendService resource](https://cloud.google.com/compute/docs/reference/v1/backendServices#resource).
+ * @param {function=} callback - The callback function.
+ * @param {?error} callback.err - An error returned while making this request.
+ * @param {module:compute/network} callback.network - The created Network
```
Comment (callmehiphop) [inline lib/compute/index.js]: Is this correct? I'm not familiar with the upstream counterpart, but if it does create a network I would think this method should be titled `createNetwork`.

Comment (stephenplusplus) [inline lib/compute/index.js]: Copy/paste error. Fixed :white_check_mark:

File: lib/compute/index.js [hunk 2]
```diff
@@ -1090,6 +1160,111 @@ Compute.prototype.getRegions = function(options, callback) {
...
+ *     - **`comparison`**: the comparison operator, `eq` (equal) or `ne`
+ *       (not equal)
+ *     - **`filterString`**: the string to filter to. For string fields, this
+ *       can be a regular expression.
+ * @param {number} options.maxResults - Maximum number of snapshots to return.
+ * @param {string} options.pageToken - A previously-returned page token
+ *     representing part of the larger set of results to view.
+ * @param {function} callback - The callback function.
+ * @param {?error} callback.err - An error returned while making this request.
+ * @param {module:compute/snapshot} callback.snapshots - Snapshot objects from
```
Comment (callmehiphop) [inline lib/compute/index.js]: I think this may be another copypasta error!

File: lib/compute/service.js [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+ */
+var ServiceObject = require('../common/service-object.js');
+
+/**
+ * @type {module:common/util}
+ * @private
+ */
+var util = require('../common/util.js');
+
+/*! Developer Documentation
```
Comment (callmehiphop) [inline lib/compute/service.js]: Looks like we forgot some documentation here :)

File: lib/compute/service.js [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+ * var gce = gcloud.compute();
+ *
+ * var service = gce.service('service-name');
+ */
+function Service(compute, name) {
+  var methods = {
+    /**
+     * Create a backend service.
+     *
+     * @param {object} config - See {module:compute#createService}.
```
Comment (callmehiphop) [inline lib/compute/service.js]: Looks like we're missing the callback param stuff

File: lib/compute/service.js [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+    /**
+     * Get a Service object if it exists.
+     *
+     * You may optionally use this to "get or create" an object by providing an
+     * object with `autoCreate` set to `true`. Any extra configuration that is
+     * normally required for the `create` method must be contained within this
+     * object as well.
+     *
+     * @param {options=} options - Configuration object.
+     * @param {boolean} options.autoCreate - Automatically create the object if
```
Comment (callmehiphop) [inline lib/compute/service.js]: Missing callback param here as well

**Comment by callmehiphop** [comment]
Is there a docs preview for this?

**Comment by stephenplusplus** [comment]
Docs preview up: http://stephenplusplus.github.io/gcloud-node/#/docs/master/compute/service

RE: the missing callback param docs, that was an intentional decision back when we introduced the famous "get or create" overhaul. While having multiple methods to get at the same data can be convenient, the consequence was having to write the *exact same documentation* in multiple places. That has some hazardous side effects if the code is updated with a docs change local to the source method, the branched methods that reference it would go out of date. The chosen path was simply to have a link back to the source method from the branched methods and let the example do a quick and dirty job of showing what will happen in your callback. If you have any ideas how we can improve on this, open an issue so we can :hash: it out!

**Comment by callmehiphop** [comment]
Thanks for the preview! I think we need to add BackendServices to the site navigation via <kbd>toc.json</kbd>!

File: docs/toc.json [hunk 6]
```diff
@@ -71,6 +71,9 @@
       "title": "Region",
       "type": "compute/region"
     }, {
+      "title": "Service",
+      "type": "compute/service"
+    }, {
```
Comment (stephenplusplus) [inline docs/toc.json]: @callmehiphop it's in here! Did I do something wrong? Also, is `implemented` still a thing?

Comment (callmehiphop) [inline docs/toc.json]: Oh weird, I don't see Service in the navigation. `implemented` went away. We version <kbd>toc.json</kbd> in the same fashion as the rest of the JSON/markdown.

**Comment by callmehiphop** [comment]
This might not be the place for it, but are there are any JSDoc tags (aside from link) that would allow us to inherit the documentation from one method to another?

File: lib/compute/service.js [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+};
+
+/**
+ * Get the most recent health check results.
+ *
+ * @resource [BackendServices: getHealth API Documentation]{@link https://cloud.google.com/compute/docs/reference/latest/backendServices/getHealth}
+ *
+ * @param {string|object} group - The fully-qualified URL of an Instance Group
+ *     resource.
+ * @param {string} - group.name - The name of the Instance Group resource.
```
Comment (callmehiphop) [inline lib/compute/service.js]: The first hyphen causes a formatting error - http://stephenplusplus.github.io/gcloud-node/#/docs/master/compute/service?method=getHealth

File: docs/toc.json (hunk 6 above)
Comment (stephenplusplus) [inline docs/toc.json]: I like it! My local "deploy preview docs" script was wrong. It's in the sidebar in the preview now: http://stephenplusplus.github.io/gcloud-node/#/docs/master/compute/service

**Comment by stephenplusplus** [comment]
Not that I know of. Possibly the callback can be a "type" that we can then reference? I'm not sure if that's real / a hack usage.

File: lib/compute/service.js (hunk 7 above)
Comment (stephenplusplus) [inline lib/compute/service.js]: Phew, fixed.

**Pull request merged**